apps/opik-backend/src/main/java/com/comet/opik/infrastructure/bi/ApplicationStartupListener.java

```diff
@@ -0,0 +1,147 @@
+package com.comet.opik.infrastructure.bi;
+
+import com.comet.opik.domain.IdGenerator;
+import com.comet.opik.infrastructure.OpikConfiguration;
+import com.comet.opik.infrastructure.lock.LockService;
+import com.google.inject.Injector;
+import jakarta.ws.rs.client.Client;
+import jakarta.ws.rs.client.ClientBuilder;
+import jakarta.ws.rs.client.Entity;
+import jakarta.ws.rs.core.MediaType;
+import jakarta.ws.rs.core.Response;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+import ru.vyarus.dropwizard.guice.module.lifecycle.GuiceyLifecycle;
+import ru.vyarus.dropwizard.guice.module.lifecycle.GuiceyLifecycleListener;
+import ru.vyarus.dropwizard.guice.module.lifecycle.event.GuiceyLifecycleEvent;
+import ru.vyarus.dropwizard.guice.module.lifecycle.event.InjectorPhaseEvent;
+
+import java.net.URI;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.comet.opik.infrastructure.lock.LockService.Lock;
+
+@Slf4j
+@RequiredArgsConstructor
+public class ApplicationStartupListener implements GuiceyLifecycleListener {
+
+    public static final String NOTIFICATION_EVENT_TYPE = "opik_os_startup_be";
+
+    // This event cannot depend on authentication
+    private final Client client = ClientBuilder.newClient();
+    private final AtomicReference<Injector> injector = new AtomicReference<>();
+
+    @Override
+    public void onEvent(GuiceyLifecycleEvent event) {
+
+        if (event instanceof InjectorPhaseEvent injectorEvent) {
+            injector.set(injectorEvent.getInjector());
+        }
+
+        if (event.getType() == GuiceyLifecycle.ApplicationStarted) {
+
+            String eventType = GuiceyLifecycle.ApplicationStarted.name();
+
+            var config = (OpikConfiguration) event.getSharedState().getConfiguration().get();
+
+            if (!config.getMetadata().getUsageReport().enabled()) {
+                log.info("Usage report is disabled");
+                return;
+            }
+
+            if (StringUtils.isEmpty(config.getMetadata().getUsageReport().url())) {
+                log.warn("Usage report URL is not set");
+                return;
+            }
+
+            var lockService = injector.get().getInstance(LockService.class);
+            var generator = injector.get().getInstance(IdGenerator.class);
+            var usageReport = injector.get().getInstance(UsageReportDAO.class);
+
+            var lock = new Lock("opik-%s".formatted(eventType));
+
+            lockService.executeWithLock(lock, tryToReportStartupEvent(usageReport, generator, eventType, config))
```

Comment:
Why a lock is required. The service starts only once. Is there any at least once guarantee instead of having a single event? Is there any other possibility of a race condition?

Reply:
We just want to trigger this event once per installation. If, for some reason, the project is being deployed as a cluster, we don't want to trigger the event multiple times, which is why we used the lock.

Reply:
Ok, that makes, although you can probably accomplish with the DB only. Anyway, this is ok to stay if the implementation is clean.